docs: create troubleshooting page (#11897)
renovatebot/renovate: docs/usage/getting-started/troubleshooting.md

# Troubleshooting Renovate

Learn how to troubleshoot problems with Renovate, where to find the logging output, and how to get help if needed.

## Getting the logs

Renovate logs the most recent jobs.
Where you can find the logs depends on how you're running Renovate.

### GitHub Hosted app

Each pull request from Renovate contains a link to the WhiteSource Renovate Dashboard in the PR body text.

The text you're looking for is:

> This PR has been generated by WhiteSource Renovate. View repository job log here.

> Click on the blue text "here" to go to the WhiteSource Renovate Dashboard.

Author: We have a "Dependency Dashboard", which shows pending dependency upgrades.
We also have another dashboard, called "WhiteSource Renovate Dashboard".

This is a bit confusing. Not sure how we can prevent the reader from getting the 2 dashboards mixed up...

Author: If all the "WhiteSource Renovate Dashboard" (WSRD) will ever do is give access to the logs, then it would be way clearer if we would call it a "log server" or something similar, as that removes any potential confusion for the end-user.

Reviewer: There's always been plans to make it much more, e.g. view extracted dependencies, request new runs, etc. It will be the "real" dashboard eventually, while the issue is more like a small shortcut version.

Author: > There's always been plans to make it much more, e.g. view extracted dependencies, request new runs, etc. It will be the "real" dashboard eventually, while the issue is more like a small shortcut version.

Oh cool! So the WSRD will be the full fat "dashboard", and the Dependency Dashboard is the lite version. 😄 

In that case maybe we should keep the terms we have now, as it won't be a log server for ever then... 😄

> Sign in with your GitHub or GitLab account.

Author: That's because GitLab still doesn't work for us as a hosted platform isn't it?

Could you use your GitLab account to access logs made on GitHub repositories? Stupid idea maybe... 😄

Maybe if you cannot really do anything functional when logged in with GitLab, maybe that option should be removed from the login screen???

> Sign in with your GitHub or GitLab account.

Reviewer: No, you can't use gitlab to access github, and I need to remove the login option soon

> Sign in with your GitHub or GitLab account.

Author: > No, you can't use gitlab to access github, and I need to remove the login option soon

Okay! 😉

> Click on the blue text "here" to go to the WhiteSource Renovate Dashboard.

Reviewer: ```suggestion
Click on the blue text "here" to go to the WhiteSource Renovate App Dashboard.
```

Once you're logged in, you can see the logs for the Renovate jobs on your repository.

Renovate only keeps the logs for the last 3 days.

When you've clicked on a recent job, you'll be able to select a debug level that you care about.
For a full overview, make sure you select the DEBUG log level.

### Selfhosted

> TODO: Explain log directory structure. Will probably be different depending on self-hosted platform...
> 
> TODO: Explain how long logs are kept by default on differnet platforms.
> 
> TODO: Explain how to toggle log level: <https://docs.renovatebot.com/examples/self-hosting/#about-the-log-level-numbers>

Reviewer: ```suggestion
The easiest way to gather logs from Renovate for any platform is to use the default logging to `stdout`/console.
By default, Renovate will log in a human-readable format at `INFO` level.

For troubleshooting it's recommended to increase logging to `DEBUG` level by adding `LOG_LEVEL=debug` to your environment variables before invoking Renovate.

If your Renovate logs are being processed by a log service before you access them, you may find it better to have Renovate output logs in JSON format instead so that they can be reliably parsed and filtered.
This can be achieved by adding `LOG_FORMAT=json` to your environment variables before invoking Renovate.
```

## Log debug levels

There are different severity levels for the log output.
From least severe to most severe:

- DEBUG
- INFO
- WARN
- ERROR
- FATAL

As you can probably imagine, the stuff that's really going wrong will likely have a level of WARN or higher.

> TODO: What do the Renovate maintainers recommend as debug level?

Author: I'm not a heavy user of the logs, Renovate bot usually does what I want anyways.
So I'm not familiar with the workflow that you as maintainers have when troubleshooting.

> As you can probably imagine, the stuff that's really going wrong will likely have a level of WARN or higher.
> 
> TODO: What do the Renovate maintainers recommend as debug level?

Reviewer: ```suggestion
To check for problems, look for `WARN` or `ERROR` logs (level 40 or 50 if in JSON format).
To troubleshoot further, you usually need to look at `DEBUG` logs.
```

> ## Relevant part of the logs

Reviewer: ```suggestion
## Resolving problems using logs
```

> Please locate the relevant parts of the logs before asking for help or posting a bug report.
> Do not expect the Renovate maintainers to read through the full logs when trying to help you, as that takes a lot of time on our part.
> 

Reviewer: ```suggestion
```

We recommend you follow this process:

1. Read the _whole_ log output, to get a feel for all the steps that Renovate is taking
1. Select - or search for - a higher debug level, like WARN, ERROR or FATAL
1. Comparing the log output from a "bad" job to the log output from a known good job

> 1. Copy/paste the relevant parts of the logs into your discussion post or bug report

Author: These steps are what I would do as a interested newcomer to reading logs, not sure if this is the most efficient way to get to the problem though. 😉

> 1. Read the _whole_ log output, to get a feel for all the steps that Renovate is taking
> 1. Select - or search for - a higher debug level, like WARN, ERROR or FATAL
> 1. Comparing the log output from a "bad" job to the log output from a known good job
> 1. Copy/paste the relevant parts of the logs into your discussion post or bug report

Reviewer: ```suggestion
1. Try to narrow in on the problem area e.g. by looking for relevant branches or WARN or ERROR messages
1. Find all relevant DEBUG or INFO messages from before and after the problem occurred
1. Copy/paste the relevant parts of the logs into your discussion post or bug report
```

> ## Getting help
> 

Reviewer: ```suggestion
```

> 1. Try to narrow in on the problem area e.g. by looking for relevant branches or WARN or ERROR messages

Author: ```suggestion
1. Try to narrow in on the problem area e.g. by looking for relevant branches or `WARN` or `ERROR` messages
```

> 1. Find all relevant DEBUG or INFO messages from before and after the problem occurred

Author: ```suggestion
1. Find all relevant `DEBUG` or `INFO` messages from before and after the problem occurred
```

> If you cannot fix the problem yourself after reading the logs, and reading - or searching through - our documentation, search the [renovatebot/renovate discussion](https://github.com/renovatebot/renovate/discussions) forum to see if somebody has asked a similar or related question.

Author: The goal of this section is to try to nudge people into helping themselves, before they're going to spend maintainer time to get their problem fixed.

> If none of these steps have helped you, then create a new discussion post to get help from the Renovate maintainers.

Reviewer: ```suggestion
If none of these steps have helped you, then create a new discussion post to get help from the Renovate maintainers.

Please locate the relevant parts of the logs as described earlier before asking for help or posting a bug report.
Do not expect the Renovate maintainers to read through the full logs when trying to help you, as that takes a lot of time on our part.
If later it turns out that the full logs are necessary, you will be asked for them then.

```
